### DISCUSSION TITLE
gettimeofday() for MSVC

### DISCUSSION THREAD

**Discussion by gvanem** [cesanta/mongoose#2674] [Feature requests]
Running the `unit_test.exe` program for `_WIN32` I noted there were no `diff: xxx` printed for the local vs. SNTP time.
I have the D4 ([Dimension 4](http://www.thinkman.com/dimension4/default.htm)) program running here which keep my PC's time rather accurate. History of last week:
![image](https://github.com/cesanta/mongoose/assets/945271/974fe9ad-88c8-45a6-9282-9088c936442c)

So I'm interested in SNTP. With the little patch below, I now got this: 
```c 
afbdaf6 3 sntp.c:45:sntp_cb             1 got time: 1711726018181 ms from epoch  (udp://time.windows.com:123)
afbdaf6 3 unit_test.c:345:sntp_cb       got time: 1711726018181
afbdaf6 3 unit_test.c:351:sntp_cb       diff: 2

afbdb24 3 sntp.c:45:sntp_cb             1 got time: 1711726018237 ms from epoch (udp://time.google.com:123)
afbdb24 3 unit_test.c:345:sntp_cb       got time: 1711726018237
afbdb24 3 unit_test.c:351:sntp_cb       diff: 8
```

Patch:
```diff
--- Git-latest/src/fs_posix.c 2024-02-13 12:00:44
+++ src/fs_posix.c            2024-03-29 16:24:49
@@ -51,8 +51,8 @@
   struct dirent result;
 } DIR;

-#if 0
-int gettimeofday(struct timeval *tv, void *tz) {
+int mg_gettimeofday(struct timeval *tv, void *tz) {
   FILETIME ft;
   unsigned __int64 tmpres = 0;

@@ -278,12 +333,17 @@
   mg_iobuf_free(&io);
 }

--- a/test/unit_test.c 2024-03-28 13:12:53
+++ b/test/unit_test.c 2024-03-29 16:26:33

+#if (MG_ARCH == MG_ARCH_WIN32)
+  int mg_gettimeofday(struct timeval *tv, void *tz);
+  #define gettimeofday(tv, tz) mg_gettimeofday (tv, tz)
+#endif
+
 static void sntp_cb(struct mg_connection *c, int ev, void *ev_data) {
   if (ev == MG_EV_SNTP_TIME) {
     int64_t received = *(int64_t *) ev_data;
     *(int64_t *) c->fn_data = received;
     MG_DEBUG(("got time: %lld", received));
-#if MG_ARCH == MG_ARCH_UNIX
+#if (MG_ARCH == MG_ARCH_UNIX) || (MG_ARCH == MG_ARCH_WIN32)
     struct timeval tv = {0, 0};
     gettimeofday(&tv, 0);
     int64_t ms = (int64_t) tv.tv_sec * 1000 + tv.tv_usec / 1000;
```

So my request is for a `mg_gettimeofday()` supporting both Windows and POSIX. How about it?

**Comment by gvanem** [comment]
*PS*. I did a similar patch for the SNTP example:
```diff
--- a/examples/sntp-time-sync/main.c 2024-01-13 14:54:39
+++ b/examples/sntp-time-sync/main.c 2024-03-29 17:01:37
@@ -22,11 +22,21 @@
   return t;
 }

+#if (MG_ARCH == MG_ARCH_WIN32)
+  int mg_gettimeofday(struct timeval *tv, void *tz);
+  #define gettimeofday(tv, tz) mg_gettimeofday (tv, tz)
+#endif
+
+
 // SNTP client callback
 static void sfn(struct mg_connection *c, int ev, void *ev_data) {
   if (ev == MG_EV_SNTP_TIME) {
     int64_t t = *(int64_t *) ev_data;
-    MG_INFO(("Got SNTP time: %lld ms from epoch", t));
+    struct timeval tv = {0, 0};
+    gettimeofday(&tv, 0);
+    int64_t ms     = (int64_t) tv.tv_sec * 1000 + tv.tv_usec / 1000;
+    int64_t t_diff = ms > t ? ms - t : t - ms;
+    MG_INFO(("Got SNTP time: %lld ms from epoch (diff: %lld)", t, t_diff));
     s_boot_timestamp = (time_t) ((t - mg_millis()) / 1000);
   } else if (ev == MG_EV_CLOSE) {
     s_sntp_conn = NULL;
```

Showing: `b1c0248 2 main.c:39:sfn     Got SNTP time: 1711728125392 ms from epoch (diff: 1)`

**Comment by scaprile** [comment]
> So my request is for a `mg_gettimeofday()` supporting both Windows and POSIX. How about it?

Current state of things is: 
- Mongoose itself does not seem to have a need for `gettimeofday`.
- `#if 0` code in `fs_posix.c` should not be enabled but probably go away, does it have an actual non-fulfilled need ?
- `gettimeofday` functionality is used only on:
   - `unit_test.c`, to show diffs, thing that we don't care much if it runs on native Windows, but we might benefit from it.
   - legacy-style (current-style) embedded examples `syscalls.c`, based on `mg_now()`, implemented on the example and based on SNTP time; apparently to feed MbedTLS:
   https://github.com/cesanta/mongoose/blob/655f5db996d2210394d98ded3339796b40e5a2b4/examples/device-dashboard/net.c#L31-L34
   https://github.com/cesanta/mongoose/blob/655f5db996d2210394d98ded3339796b40e5a2b4/examples/stm32/nucleo-f746zg-make-baremetal-builtin/syscalls.c#L90-L98

So... `mg_gettimeofday` could serve the need for `gettimeofday` in Windows and embedded hardware when working with MbedTLS ?

**Reply by gvanem** [reply]
A `mg_gettimeofday()` should return the Unix-time. Not some time based on boot-time.

**Reply by cpq** [reply]
Currently we have this:

1. mg_millis() is a monotonic timer increasing each millisecond. On windows and embedded system, this is milliseconds since boot. On unix, that's some weird number that increases every millisecond
2.  We have mg_now() prototype defined in the header, but not implemented in the core. This function supposed to return a number of milliseconds since epoch. Kinda like mg_gettimeofday, but with a simpler interface. We don't have  it implemented in the core because it needs SNTP. And SNTP may not be used at all by the user application

@gvanem does mg_now() seem like the function you need?
